lib/go-atscfg/remapdotconfig.go

```diff
@@ -0,0 +1,413 @@
+			}
+			midRemap += ` @plugin=cachekey.so`
+			for name, val := range profilesCacheKeyConfigParams[*ds.ProfileID] {
+				midRemap += ` @pparam=--` + name + "=" + val
+			}
+		}
+		if ds.RangeRequestHandling != nil && *ds.RangeRequestHandling == tc.RangeRequestHandlingCacheRangeRequest {
+			midRemap += ` @plugin=cache_range_requests.so`
+		}
+
+		if midRemap != "" {
+			midRemaps[*ds.OriginFQDN] = midRemap
+		}
+	}
+
+	textLines := []string{}
+	for originFQDN, midRemap := range midRemaps {
+		textLines = append(textLines, "map "+originFQDN+" "+originFQDN+midRemap+"\n")
+	}
+	sort.Strings(textLines)
+
+	text := header
+	for _, line := range textLines {
+		text += line
+	}
+	return text
+}
+
+func GetServerConfigRemapDotConfigForEdge(
+	cacheURLConfigParams map[string]string,
+	profilesCacheKeyConfigParams map[int]map[string]string,
+	serverPackageParamData map[string]string, // map[paramName]paramVal for this server, config file 'package'
+	server *ServerInfo,
+	dses []RemapConfigDSData,
+	atsMajorVersion int,
+	header string,
+) string {
+	for name, val := range serverPackageParamData {
+		serverPackageParamData[name] = strings.Replace(val, "__HOSTNAME__", server.HostName+"."+server.DomainName, -1)
+	}
+
+	textLines := []string{}
+
+	for _, ds := range dses {
+		remapText := ""
+		if ds.Type == tc.DSTypeAnyMap {
+			if ds.RemapText == nil {
+				log.Errorln("ds '" + ds.Name + "' is ANY_MAP, but has no remap text - skipping")
+				continue
+			}
+			remapText = *ds.RemapText + "\n"
+			textLines = append(textLines, remapText)
+			continue
+		}
+
+		remapLines, err := MakeEdgeDSDataRemapLines(ds, server)
+		if err != nil {
+			log.Errorln("making remap lines for DS '" + ds.Name + "' - skipping! : " + err.Error())
+			continue
+		}
+
+		for _, line := range remapLines {
+			profilecacheKeyConfigParams := (map[string]string)(nil)
+			if ds.ProfileID != nil {
+				profilecacheKeyConfigParams = profilesCacheKeyConfigParams[*ds.ProfileID]
+			}
+			remapText = BuildRemapLine(cacheURLConfigParams, atsMajorVersion, server, serverPackageParamData, remapText, ds, line.From, line.To, profilecacheKeyConfigParams)
+		}
+		textLines = append(textLines, remapText)
+	}
+
+	text := header
+	sort.Sort(sort.StringSlice(textLines))
+	for _, line := range textLines {
+		text += line
+	}
+	return text
+}
+
+// BuildRemapLine builds the remap line for the given server and delivery service.
+// The cacheKeyConfigParams map may be nil, if this ds profile had no cache key config params.
+func BuildRemapLine(cacheURLConfigParams map[string]string, atsMajorVersion int, server *ServerInfo, pData map[string]string, text string, ds RemapConfigDSData, mapFrom string, mapTo string, cacheKeyConfigParams map[string]string) string {
+	// ds = 'remap' in perl
+
+	mapFrom = strings.Replace(mapFrom, `__http__`, server.HostName, -1)
```

Comment:
`__http__` resolves to the hostname? If this isn't a typo, it's undocumented.

Reply:
This is transliterating Perl: https://github.com/apache/trafficcontrol/blob/56ef95e1783360b4afb600b06cf138255fb34c89/traffic_ops/app/lib/API/Configs/ApacheTrafficServer.pm#L2864

Reply:
*sigh* - well then it'll need to get documented. I'll open an issue.